components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -273,18 +286,41 @@ open class GleanInternalAPI internal constructor () {
             return
         }
 
+        if (!uploadEnabled) {
+            logger.error("Glean must be enabled before handling events.")
+            return
+        }
+
+        var availableToSend = false
+
         when (pingEvent) {
             Glean.PingEvent.Background -> {
-                sendPing(BaselinePing.STORE_NAME, BaselinePing.STORE_NAME)
-                sendPing("events", "events")
+                // Schedule the baseline and event pings and save the result to determine whether or not
+                // we need to schedule the upload worker
+                assembleAndSerializePing(BaselinePing.STORE_NAME)
+                assembleAndSerializePing("events")
+
+                // Always true since baseline is always sent
+                availableToSend = true
             }
             Glean.PingEvent.Default -> {
+                // Check the metrics ping schedule to determine whether it's time to schedule a new
+                // metrics ping or not
                 if (metricsPingScheduler.canSendPing()) {
-                    sendPing(MetricsPingScheduler.STORE_NAME, MetricsPingScheduler.STORE_NAME)
+                    availableToSend = assembleAndSerializePing(MetricsPingScheduler.STORE_NAME)
                     metricsPingScheduler.updateSentTimestamp()
                 }
             }
         }
```

Comment:
nit: You can also do assignments from `when` statements. Up to you if that is more readable here. :)

```kotlin
val availableToSend = when (pingEvent) {
    Glean.PingEvent.Background -> {
       ...
       true
    }
    Glean.PingEvent.Default -> {
         if (metricsPingScheduler.canSendPing()) {
             ...
             true
         } else {
             false
         }
    }
}
```

Reply:
That's awesome,  I like it!